docs/arch/002-store-responses-in-query-string.md

```diff
@@ -0,0 +1,61 @@
+# ADR 2: Store responses in query string
+
+## Background
+
+At the time of writing Smart Answer's has two methods of storing user responses: appending to the URL path or saving to the user's session cookie.
+
+The original and most used method is appending responses to the URL path, for example `gov.uk/flow-name/y/response1/response2/response3`. This generates an array of responses that Smart Answers uses to resolve which question or outcome needs to be displayed.
+
+The other method is storing user's responses in session cookies. This functionality was added when we added the `find-coronavirus-support` flow. This flow handled sensitive information that needed to be hidden from the URL. The use of sessions was also required to support the `Leave this site` component, which allowed users to quick erase their responses and prevent them from being recovered. The user responses are stored as a hash keyed by the name of the node (question) which provided the response, for example:
+
+```ruby
+session = {
+  'flow-name': {
+     'question1': 'response1',
+     'question2': 'response2',
+     'question3': 'response3',
+     }
+}
+```
+
+When Smart Answer's was originally built the use of session was thought to be unnecessary as user data expected to be handled wasn't sensitive. The reasoning for appending to the URL path instead of using a query string is unclear.
+
+## Problem
+
+There are several drawbacks from using the URL path to store user responses:
+
+1. There is no information within the URL to indicate which page the user is requesting. This makes it impossible to understand which page is being displayed to the user, without knowing internal logic of a particular flow. This is problematic for understanding user analytics, request logs and debugging. Our current user analytics determine which page is displayed to a user from content within the page (e.g. heading text). This is fragile as content changes and cumbersome as we have to update our analytics logic when we change the content. As for request logs, there is no easy way to calculate metrics per page as multiple request URLs may resolve to the same page.
+
+1. It is difficult to generate URLs with pre-populate responses, for example `/flow-name/y/response1/response2/<pre-populated-response>`. Unless the response is for the first question in a flow, it requires pre-filling responses for all former questions which are unlikely to be known. 
+
+1. Changes to a flow can cause issues with previously cached pages. As URLs are not deterministic for a particular page, changes to a flow can alter which page is displayed. For example, a given flow displays an outcome page for the path `/x/y/z`. A question is removed and the flow now shows an outcome page for the path `/x/y`. The cache would continue to return the old question and users would then be redirected to `/x/y/z`, which is no longer a valid path and could result in an error if it was not also cached.
```

Comment:
Might be worth nothing this situation also happens when the questions are not moved, but the answer is changed.  For example: `/x/y/z` - becomes `/x/y/a` where `z` was `Yes` and `a` is `Yes, but...`

Reply:
Oh yeah - that too! I think there are probably lots of weird caching situation that this may help with. I hope the general gist comes across that changes to  cached flows can cause problems.